### PR TITLE
MHC Projection Fusion

### PR DESCRIPTION
**Pull request by muskansh-google**
# Description

This PR optimizes the Manifold-Constrained Hyper Connections (mHC) layer by fusing three independent projection matmuls (pre, post, and residual) into a single unified projection to reduce TPU kernel launch overhead.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.
- Manually tested for a custom model on v6e-256 chips and got a 0.3% imrpovement in TFLOPs
- Updated test_mhc_lite_doubly_stochastic to work with the updated post-matmul API. 
- Added test_weight_concatenation_equivalence to verify mathematical equivalence between the fused projection and the sequential baseline.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by dandragona — APPROVED** [review]
LGTM. Thank you!

File: src/maxtext/layers/mhc.py [hunk 1]
```diff
@@ -214,22 +210,16 @@ def res_mapping(self, x: Array):
...
-      h_res = jnp.einsum("bsm,mn -> bsn", x, res_alpha, precision=self.matmul_precision)
       b, s, _ = h_res.shape
       h_res = jnp.reshape(h_res, (b, s, self.k, self.k))
       intermediate = res_alpha_scale * h_res + res_beta[None, None, :, :]
       output = sinkhorn(intermediate, self.sinkhorn_iterations)
       return output
 
-  def mapping(self, x: Array, alpha_scale: Array, alpha: Array, beta: Array, scale: float, eps: float = 0.0):
-    """Helper function for both pre and post mappings."""
-    # In MaxText, we match weight precision to activations before Matmul
```
Comment (RissyRan) [inline src/maxtext/layers/mhc.py:227]: Shall we keep this comment? It should also apply to beta and alpha_scale.

File: src/maxtext/layers/mhc.py [hunk 2]
```diff
@@ -191,16 +191,12 @@ def __init__(
         out_sharding=(None,),
     )
 
-  def res_mapping(self, x: Array):
-    """Helper function for residual mapping."""
-    # In MaxText, we match weight precision to activations before Matmul
```
Comment (RissyRan) [inline src/maxtext/layers/mhc.py:196]: How we keep this comment?

**Review by RissyRan — APPROVED** [review]
LGTM, just a minor comment.

File: src/maxtext/layers/mhc.py (hunk 2 above)
Comment (muskansh-google) [reply src/maxtext/layers/mhc.py:196]: Done. Thank you!
